### DISCUSSION TITLE
Autogen for data analysis including numbers and text?

### DISCUSSION THREAD

**Discussion by analyticsinsights** [microsoft/autogen#4726] [Q&A] [answered]
I would like to use Autogen to analyse data stored in a CSV-file. When I set this up as a [GroupChat](https://microsoft.github.io/autogen/0.2/docs/notebooks/agentchat_groupchat/), I am able to have, for example, a (1) DataLoader, (2) Coder and (3) CodeExecutor agent which are together able to write and execute a Python code which create a chart that show basic statistics of the data in the CSV-file. However, in the CSV-file are also text data in each row and I would like Autogen to summarise the text in each row briefly, or classify the text into topics using the LLM and then store the results back into a pandas dataframe and then work with the new content of the dataframe to create graphs.

Unfortunately, I am not able to create an agent that is able to use the LLM for this task and apply it to filtered rows of the dataframe. Somehow, the coder agent mentioned above tries instead to install NLTK or Scikit-Learn and count words of the text instead of simply using the LLM to summarise the text. How can I achieve this? 

[This example](https://github.com/sugarforever/AutoGen-Tutorials/blob/main/autogen_rag_agent.ipynb) of using RAG seems to allow a specific file with data, but I would need the data to be preprocessed in interim steps by the other agents (as described above) and only the filtered Pandas dataframe should then be used by the RetrieveUserProxyAgent.

**Comment by rysweet** [comment]
@analyticsinsights prompting for coding tasks can be tricky. unfortunately without an example of your code to look at I think its hard to advise you on next steps.

**Reply by analyticsinsights** [reply]
@rysweet Here is an example code. 

The idea is that the agents are able (based on the specific instructions in the initiate_chat message) to filter the data and then use a LLM to work with the text that is included this filtered dataframe.

```
import os
import autogen
from autogen.coding import LocalCommandLineCodeExecutor

# Specify the model to be used
config_list = [
    {
        "model": "gpt-4o-mini",
        "api_key": os.environ["OPENAI_API_KEY"]
    }
]

# Define the file path to your CSV
file_path = 'https://raw.githubusercontent.com/LearnDataSci/articles/refs/heads/master/Python%20Pandas%20Tutorial%20A%20Complete%20Introduction%20for%20Beginners/IMDB-Movie-Data.csv'

# Create the Manager Agent
manager = autogen.AssistantAgent(
    name="Manager",
    llm_config={"config_list": config_list},
    system_message="You are the Manager Agent responsible for coordinating the workflow among agents to accomplish diverse data analysis tasks. Your role includes understanding the user's objectives, delegating responsibilities to appropriate agents, and ensuring the seamless execution of tasks to meet user requirements.",
)

# Create the Data Loader Agent
data_loader = autogen.AssistantAgent(
    name="Data_Loader",
    llm_config={"config_list": config_list},
    system_message="You are the Data Loader Agent. Your primary responsibility is to load and preprocess data from specified sources, such as CSV files, databases, or APIs. You should extract relevant information as per the task requirements and prepare it for analysis by other agents.",
)

# Create the Code Executor Agent
code_executor = autogen.UserProxyAgent(
    name="Code_Executor",
    human_input_mode="NEVER",
    max_consecutive_auto_reply=20,
    code_execution_config={
        "executor": LocalCommandLineCodeExecutor(work_dir="work_dir"),
    },
    system_message="You are the Code Executor Agent. Your role is to execute code provided by other agents to process and analyze data. Ensure that the code runs efficiently and correctly, handling any errors that may arise during execution, and provide the results back to the requesting agents.",
)

# Create the Analysis Agent
analysis_agent = autogen.AssistantAgent(
    name="Analysis_Agent",
    llm_config={"config_list": config_list},
    system_message="You are the Analysis Agent. Your primary responsibility is to analyze provided textual data to extract and summarize the main themes, insights, or patterns as specified by the task requirements. Focus on delivering concise and coherent thematic summaries.",
)

# Set up the Group Chat
groupchat = autogen.GroupChat(
    agents=[manager, data_loader, code_executor, analysis_agent],
    messages=[],
    max_round=10,
)

# Initialize the Group Chat Manager
manager_agent = autogen.GroupChatManager(
    groupchat=groupchat,
    llm_config={"config_list": config_list},
    is_termination_msg=lambda msg: "terminate" in msg["content"].lower()
)

# Start the conversation with the Manager Agent initiating the task
manager.initiate_chat(
    manager_agent,
    message=f"Based on the data in {file_path}, provide a brief summary for each of Ridley Scott's movies. Compile these thematic summaries into a markdown file named 'movie_summaries.md'",
    summary_method="reflection_with_llm",
    clear_history=True
)
```

**Reply by ekzhu** [reply]
You need to set the `description` field for the agents to make then work in a group chat.

Also, I suggest let the manager playing the role of planner.

I recommend using the 0.4 API for this. We will be rolling out the official upgrade in the next few weeks. 

Generally, group chat isn't going to be very stable and robust. If the workflow can be described ahead of time it is better to use a sequential workflow approach.

**Reply by analyticsinsights** [reply]
@ekzhu Many thanks for your reply. 

I have adapted my code above based on [this example of the Multi-Agent Group Chat](https://microsoft.github.io/autogen/0.2/docs/notebooks/agentchat_groupchat_research) from the official documentation. I don't see that a `description` field has been used there. Hence, I also wouldn't know what to put there concretely to make it work.

Maybe waiting for Autogen 0.4 is the solution then, but any concrete suggestion with a coding example would be appreciated. I am looking for an example that actually analyses data that comes in the form of a CSV-file, that can execute Python code to create e.g. graphs and can also use LLMs to deal with text fields.

**Reply by ekzhu** [reply]
@analyticsinsights in that example the description field would be helpful.

I think you can start with a single agent and built tools. 

Tools -- cheaper, fast, robust, explainable
LLM -- expensive, slow, unreliable, unexplainable 

Use tools and orchestrated workflows to reduce as much stochasticity as possible. 

Only go multi agent when you have exhausted the capability of single agent.

**Reply by analyticsinsights** [reply]
Unfortunately, this is not too helpful. 

I know what tools are, and I have used a single agent at the beginning which did not get me further. It is not apparent from the [description of tool use](https://microsoft.github.io/autogen/0.2/docs/tutorial/tool-use/) which elements of my described workflow I need to embed in custom tools.

I found [this more elaborate example code](https://microsoft.github.io/autogen/0.2/docs/notebooks/agentchat_function_call_code_writing/) and will take a closer look at what functions are defined and how. I suppose that I need to implement (1) loading of the CSV-file, (2) filtering of the CSV-file and (3) calling LLM for summarisation as functions. But I am concerned that the final tool becomes very static if I am pre-defining all these elements. Then the tool would lack flexibility, which is precisely one of the key advantages that agentic workflows promised to provide.

What is also needed with Autogen is a list of standard tools like [CrewAI](https://docs.crewai.com/concepts/tools) and [LangChain](https://python.langchain.com/docs/integrations/tools/) have it - and to document them and their application well.

**Reply by ekzhu** [reply]
Looks like you have a good sense of where you want to go. My impression was that you had a fixed workflow so my suggestion was to apply determinism and no need to worry about generalization. Apologies. 

If you want your agent to generalize to a wide range of tasks, code executor or functions that writes code are helpful. But they come with the additional cost and risk. 

You can use langchain tools through a wrapper: https://microsoft.github.io/autogen/dev/reference/python/autogen_ext.tools.langchain.html

**Reply by ekzhu** [reply]
Our work on general capability is MagenticOne. It will be available in the 0.4 version. 

Are you prototyping or looking for something to work in production scenarios?

**Reply by analyticsinsights** [reply]
This is for prototyping, I am trying to implement various use-cases. Analysing data stored in CSV-files is clearly one of them.

**Comment by analyticsinsights** [comment]
I just wanted to share with the community that answering questions on the content of a CSV-file is actually very straightforward --out of the box-- with [LangChain tools](https://python.langchain.com/docs/integrations/tools/pandas/).

```
import pandas as pd
from langchain_experimental.agents.agent_toolkits import create_pandas_dataframe_agent
from langchain_openai import OpenAI

df = pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv")
agent = create_pandas_dataframe_agent(OpenAI(temperature=0), df, verbose=True, allow_dangerous_code=True)
agent.invoke("What is the percentage of passengers that were in first class?")
```

The [AG2 release](https://docs.ag2.ai/blog/2024-12-20-Tools-interoperability/index) to integrate tools from third-party providers shows how important tools are for agentic workflows. It would be great to have this also in Autogen.

**Reply by ekzhu** [reply]
How about this?

```python
import asyncio
import pandas as pd
from langchain_experimental.tools.python.tool import PythonAstREPLTool
from autogen_ext.tools.langchain import LangChainToolAdapter
from autogen_ext.models.openai import OpenAIChatCompletionClient
from autogen_agentchat.agents import AssistantAgent
from autogen_agentchat.ui import Console

async def main() -> None:
    df = pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv")
    tool = LangChainToolAdapter(PythonAstREPLTool(locals={"df": df}))
    model_client = OpenAIChatCompletionClient(model="gpt-4o")
    agent = AssistantAgent("assistant", tools=[tool], model_client=model_client, system_message="Use the `df` variable to access the dataset.")
    await Console(agent.run_stream(task="What's the average age of the passengers?"))

asyncio.run(main())
```

We already support Langchain tool integration through LangchainToolWrapper: https://microsoft.github.io/autogen/dev/reference/python/autogen_ext.tools.langchain.html

**Reply by analyticsinsights** [reply]
Many thanks @ekzhu! This works. 

Is it possible to return the dataframe when my task is "Filter the dataframe for first class passengers"?

**Reply by ekzhu** [reply, marked as answer]
Unless you want to convert the returned dataframe in text message, I would use a separate variable to store the result or saving the dataframe on disk. 

```python
import asyncio
import pandas as pd
from langchain_experimental.tools.python.tool import PythonAstREPLTool
from autogen_ext.tools.langchain import LangChainToolAdapter
from autogen_ext.models.openai import OpenAIChatCompletionClient
from autogen_agentchat.agents import AssistantAgent
from autogen_agentchat.ui import Console

async def main() -> None:
    df = pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv")
    repl = PythonAstREPLTool(locals={"df": df})
    tool = LangChainToolAdapter(repl)
    model_client = OpenAIChatCompletionClient(model="gpt-4o")
    agent = AssistantAgent("assistant", tools=[tool], model_client=model_client, system_message="Use the `df` variable to access the dataset.")
    await Console(agent.run_stream(task="Filter the dataframe for first class passengers and save it to a variable named `first_class`."))
    print(repl.locals.get("first_class"))

asyncio.run(main())
```

**Reply by analyticsinsights** [reply]
Many thanks @ekzhu! With this code one can really start analysing pandas dataframes and get the results using Autogen.
